File: OsmAndWear/src/net/osmand/plus/settings/backend/preferences/OsmandPreference.java
package net.osmand.plus.settings.backend.preferences;

import net.osmand.StateChangedListener;
import net.osmand.plus.settings.backend.ApplicationMode;

import org.json.JSONException;
import org.json.JSONObject;

public interface OsmandPreference<T> {
	T get();

	boolean set(T obj);

	boolean setModeValue(ApplicationMode m, T obj);

	T getModeValue(ApplicationMode m);

	String getId();

	void resetToDefault();

	void resetModeToDefault(ApplicationMode m);

	void overrideDefaultValue(T newDefaultValue);

	void addListener(StateChangedListener<T> listener);

	void removeListener(StateChangedListener<T> listener);

	boolean isSet();

	boolean isSetForMode(ApplicationMode m);

	boolean writeToJson(JSONObject json, ApplicationMode appMode) throws JSONException;

	void readFromJson(JSONObject json, ApplicationMode appMode) throws JSONException;

	String asString();

	String asStringModeValue(ApplicationMode m);

	T parseString(String s);
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/settings/backend/preferences/OsmandPreference.kt
package net.osmand.shared.settings.backend.preferences

import kotlinx.serialization.json.JsonObject
import net.osmand.shared.api.KStateChangedListener
import net.osmand.shared.settings.backend.ApplicationMode

interface OsmandPreference<T> {
	fun get(): T

	fun set(obj: T): Boolean

	fun setModeValue(m: ApplicationMode, obj: T): Boolean

	fun getModeValue(m: ApplicationMode): T

	fun getId(): String

	fun resetToDefault()

	fun resetModeToDefault(m: ApplicationMode)

	fun overrideDefaultValue(newDefaultValue: T)

	fun addListener(listener: KStateChangedListener<T>)

	fun removeListener(listener: KStateChangedListener<T>)

	fun isSet(): Boolean

	fun isSetForMode(m: ApplicationMode): Boolean

	fun writeToJson(json: JsonObject, appMode: ApplicationMode): Boolean

	fun readFromJson(json: JsonObject, appMode: ApplicationMode): Boolean

	fun asString(): String

	fun asStringModeValue(m: ApplicationMode): String

	fun parseString(s: String): T
}